### ISSUE TITLE
All results are empty arrays while the data is actually in the database

### ISSUE THREAD

**Issue by TheImpressionist** [Automattic/mongoose#7348]
**Do you want to request a *feature* or report a *bug*?**

**What is the current behavior?**

Mongoose doesn't seem to connect to the database, yet fires the `connected` event.
Everything seems to work fine, until you try to pull data. It just returns empty arrays.

This basically happened over night, yet no changes were made.

I can connect to the database through the terminal or Studio 3T just fine, and the data is all there.

```typescript
    const query: any = { email };
    const user: Array<IUserModel> = await User
      .find(query, null, { collation: { locale: 'en', strength: 2 } })
      .lean()
      .limit(1)
      .select('_id email password salt');

    return user[0];
```

```javascript
const user = await User.find({});
```

`user` here will be an empty array.

NodeJS - 11.2.0
MongoDB - 4.0
Mongoose - 5.3.0

**Comment by mhombach** [comment]
Please, for debugging thi issue, try to use `mongoose.set('bufferCommands', false);` to prevent CommandBuffering and give us the code where you connect to the database :)

**Comment by TheImpressionist** [comment]
```typescript
import * as mongoose from 'mongoose';

import { info } from 'winston';

mongoose.set('debug', true);
mongoose.set('bufferCommands', false);

export async function connect(): Promise<void> {
  try {
    if (mongoose.connection.readyState > 0 && mongoose.connection.readyState !== 0) {
      return;
    }

    const auth = process.env.NODE_ENV !== 'development'
      ? {
        auth: { authdb: process.env.DB_OPTIONS_AUTH_DB },
        user: process.env.DB_OPTIONS_USER,
        pass: process.env.DB_OPTIONS_PASS,
      }
      : {};
    const options = {
      ...auth,
      // db: { native_parser: process.env.DB_OPTIONS_DB_NATIVE_PARSER },
      useNewUrlParser: true,
      keepAlive: 1,
    };
    await mongoose.connect(process.env.DB_URI, options);
    info(':::DB CONNECTED:::');
  } catch (err) {
    throw err;
  }
}
```

This would be the connection function.

I've noticed that you can still create users and have the data persist.
It seems to be connecting to the right database, but not finding anything there, other than new entries.
I try to connect to it using Studio 3T, and all the old data is there, but not the new one.

So I figure mongoose is connecting to some ethereal plane and keeping my data safe there. Which is odd. Because the connection string and the DB URL is correct and hasn't changed in months.

**Comment by TheImpressionist** [comment]
Just tried connecting to another one of my databases which contain same data. It now cannot recognise the users in that database either. Basically, I cannot login with legit details. But I can create entries with same emails.

I presume mongoose is connecting somewhere it really shouldn't?

**Comment by mhombach** [comment]
Can you run `mongoose.connection.db.command({serverStatus: 1});` and look into the output and if `host` is what you would expect?

**Comment by TheImpressionist** [comment]
I get the following as host: `host: '98a455c8dc1f'`,

Though there's an odd thing just underneath it.

`version: '4.0.4'` - When my DB version is `4.0.2`

I see no other processes running. Really weird.

**Comment by mhombach** [comment]
Log:
- mongoose.connection.host
- mongoose.connection.port
- mongoose.connection.name

this should give you the detailed information about the used connection, maybe this helps?

**Comment by TheImpressionist** [comment]
All the details there are correct.

I guess I'm gonna have to live with it. Wouldn't be bad if I could access this mysterious database through command line or Studio 3T. But it's nowhere to be found.

I'll try downgrading mongoose version. See if that helps.

**Comment by mhombach** [comment]
@TheImpressionist

> I guess I'm gonna have to live with it.

I guess/hope that's a joke.

- You have told us that the data are NOT saved within you real database and that you have 100% verified that it's not just a typo in a model and you are looking at thr wrong collection.
- You have told us that the data saved **persist** (not only within a mongoose-session but also after, for example, a restart of the computer. So the MUST be saved somewhere..
- You have told us that you disbled `bufferCommands` so that leaves mongoose no other choice but to save the data in a REAL database.

Summed up: If all above is true, then you have an online database where those data are stored. I would not only search the problem in relation to mongoose but to find this 2nd database.
- Check the exact writing of the host/port(database
- Insert some special testdata liek a string that won't appear somewhere else and then search-query your whole database for it
- Write down (screenshot) the used storage of your database, then insert like 100-500 MB of new data into that "ghost-db". While doing that, monitor your local ram-usage to see if it, for whatever reason, is written to your ram.
- Disconnect your internet connection and try if you still can write to that ghost-db.
- Check the logfiles of your online db and see what entries there maybe a hint for what's happening.

I am personally interested in what's going on but also we need to 100% make sure that this isn't some bug in mongoose that stores data in some temporary file on harddisk even if buffering is disabled.

Looking forward to your analysis-report :)

**Comment by TheImpressionist** [comment]
Thanks, I'll go through those steps. I'm really intrigued as well.

**Comment by mhombach** [comment]
Just for clarification, you are coding on a PC and your DB is running on your laptop? Asking because of `"Did a shutdown on my laptop, now it doesn't connect"` seems like you are using such a scenario.

**Comment by TheImpressionist** [comment]
I phrased it wrong.

After the shutdown and startup it would not connect. Everything is running locally.

**Comment by mhombach** [comment]
Ok... so after a clean shutdown and booting up again, it doesn't connect... Try, step by step, to start your environment as usual and after each step check if mongoose connects.

**Comment by TheImpressionist** [comment]
Alright, I tried connecting to the DB without the `mongod` process running, but using Studio 3T.

It actually connects, and shows these mysterious databases.
So I found the data.

Now figure this. It was connecting to a docker container which was running another mongodb database from another project.

Odd thing about it really, because the container would start up itself automatically, without no one asking.

After shutting them down, now I've access to my good ol' data again.

Summed up, this is not a mongoose issue, but rather some black magic which made docker start up containers for whatever reason.

Closing this.
